### ISSUE TITLE
jid undefined inside batch success

### ISSUE THREAD

**Issue by homanchou** [mperham/sidekiq#2178]
jid method seems to be nil inside on_success "workers", is there another way to get it?

**Comment by mperham** [comment]
I don't understand the issue.  A batch's success callback is independent of any of the jobs within the batch, right?  Do you want the JID of the callback job itself?

**Comment by homanchou** [comment]
Yes.  That's exactly what I want.  Long story short: We have complex jobs which go through several stages, sometimes calling other workers and batching out, returning, repeating, etc.  Sometimes jobs disappear or the process never finishes, and there can be many reasons for this (but without more logging I'll never figure out why exactly).  And some jobs are continually kicked in by a cron, but we'd like to avoid restarting long jobs that may still be worked on.  To this end we are employing crude timestaps on an active record row before starting any new long running job, along with jid, or bid if spawned.  We use sidekiq batch status to query by bid, as well as sidekiq status gem to query by jid to see if a job is still :pending or being worked on.  We store the jid, or bid in active record which the cron can first query the jid/bid status before launching a new job.  The only (maybe not only) problem I see is that when returning from a batch, and the execution is now in the callback job itself, the status of bid is considered :completed, so if that worker is taking a long time it is eligible to be restarted.  I want to get the jid of the callback job itself and store that in active record asap so that a query via jid will show that the job is still :pending.  (I know this is not a perfect solution, and is still susceptible to race conditions but it's better than what we have now, I've tried using unique_jobs but that's even more confusing as jobs can disappear or not retry based on matching args, so I've just resorted to logging more events in activerecord so at least I can see what's going on).

**Comment by mperham** [comment]
Ok, let me think about it and see if I can't make this possible for the 2.0 release.

For now, you could roll your own middleware which stores the JID for any callback job in a thread local and access it from your callback code that way.

**Comment by mperham** [comment]
Easy fix, starting in 2.0 you will be able to add a `jid` accessor to your callback class and Pro will set the attribute before executing the callback.

```ruby
  class MyCallback
    attr_accessor :jid
    def on_complete(status, options)
    end
    def on_success(status, options)
    end
  end
```
